File: examples/mercs2/ess_probe.js
#!/usr/bin/env node
// Phase-0 spike for the Ess.UI rework: does GFx 2.1.57 in Mercenaries 2 actually
// support the runtime creation + drawing API, and can a runtime-created text
// field resolve the movie's IMPORTED font?
//
// The SDK source says yes to all of it (createEmptyMovieClip, createTextField,
// beginFill/curveTo/endFill, TextFormat, attachMovie, getNextHighestDepth,
// removeMovieClip are all on GFxSprite's method table). This confirms it in the
// real engine, because "the SDK has the method" and "it works in this build with
// this movie's font setup" are different claims.
//
// There is no way to see the game's screen from the harness, so nothing here
// relies on looking at pixels. Every check reports a NUMBER or a flag that would
// be impossible to produce if the feature were broken:
//
//   * drawing         -> getBounds() on the drawn clip. Non-zero extents mean
//                        real geometry landed in the display list.
//   * font resolution -> textWidth after setting text + TextFormat. A positive
//                        width means glyphs were found and measured; 0 means the
//                        font did not resolve, which is the failure that would
//                        force the design back to pooled attachMovie rows.
//   * an authored field's textWidth is reported alongside as a CONTROL, so a
//                        zero can be attributed to "runtime fields can't get the
//                        font" rather than "this movie has no font at all".
//
// Build:  node examples/mercs2/ess_probe.js   -> examples/mercs2/ess_probe.gfx
// Drive:  examples/mercs2/ess_probe.lua       (same directory)
//
// SHARED CONTRACT with ess_probe.lua — both sides must agree:
//   asset name      "ess_probe"        (inject as this)
//   Lua -> movie    RunProbe()         runs every check, returns + fires the event
//   movie -> Lua    fscommand("essProbe", "<k=v;k=v;...>")
//   exported symbol "ProbeRow"         attachMovie target
//   authored field  instance "ctl"     the font control

const path = require('path');
const fs = require('fs');

// The codec is browser-style (IIFE + globals), so load it through the repo's own
// test harness rather than re-implementing the sandbox — same approach as
// battery_test.js, and it exercises the real source files.
const { loadContext } = require(path.join(__dirname, '..', '..', 'tests', 'run.js'));
const { Compiler, GFMovie, Verify } = loadContext();

// ---------------------------------------------------------------- the AS2 probe
//
// Deliberately written against only the compiler's supported subset. Every
// result is stringified into one compact line because the live-test harness
// reads results out of a log file and truncates at the first newline.
const AS2 = `
// Runs every capability check and reports one flat "k=v;k=v;..." line.
// Called from Lua rather than running on frame 1, so the host's event handler is
// certainly registered before anything fires.
function RunProbe() {
    var r = "";
    var root = _root;

    // --- 1. runtime clip creation -------------------------------------------
    var holder = root.createEmptyMovieClip("probeHolder", 500);
    r = r + "cec=" + (holder == undefined ? "0" : "1") + ";";

    // --- 2. the vector drawing API ------------------------------------------
    // Includes curveTo, since rounded chrome is the whole point of drawing
    // procedurally rather than stretching authored art.
    holder.lineStyle(1, 0x59D0FF, 100);
    holder.beginFill(0x1A1C22, 90);
    holder.moveTo(0, 0);
    holder.lineTo(120, 0);
    holder.lineTo(120, 34);
    holder.curveTo(120, 40, 114, 40);
    holder.lineTo(0, 40);
    holder.lineTo(0, 0);
    holder.endFill();

    // getBounds is the only observable proof geometry landed -- no screenshots.
    var b = holder.getBounds(root);
    if (b == undefined) {
        r = r + "bw=nil;bh=nil;";
    } else {
        r = r + "bw=" + Math.round(b.xMax - b.xMin) + ";bh=" + Math.round(b.yMax - b.yMin) + ";";
    }

    // --- 3. gradient fill ----------------------------------------------------
    // Reported separately: a gradient failing while a solid fill works would
    // only cost the theme its gradients, not the design.
    var g = root.createEmptyMovieClip("probeGrad", 501);
    var okg = 1;
    g.beginGradientFill("linear", [0x59D0FF, 0x1A1C22], [100, 100], [0, 255]);
    g.moveTo(0, 0); g.lineTo(60, 0); g.lineTo(60, 20); g.lineTo(0, 20); g.lineTo(0, 0);
    g.endFill();
    var gb = g.getBounds(root);
    if (gb == undefined) { okg = 0; } else { if (gb.xMax - gb.xMin < 1) { okg = 0; } }
    r = r + "grad=" + okg + ";";

    // --- 4. runtime text field + the IMPORTED font --------------------------
    // The load-bearing check. textWidth > 0 means the font resolved.
    holder.createTextField("t1", 1, 4, 4, 110, 18);
    var t = holder.t1;
    r = r + "ctf=" + (t == undefined ? "0" : "1") + ";";
    if (t != undefined) {
        var fmt = new TextFormat();
        fmt.font = "_normal_Font";
        fmt.size = 13;
        fmt.color = 0xE1E5EC;
        r = r + "tfm=" + (fmt == undefined ? "0" : "1") + ";";
        t.text = "PROBE0123";
        t.setTextFormat(fmt);
        r = r + "tw=" + Math.round(t.textWidth) + ";";
        // Retry with embedFonts, which is what actually forces an imported
        // (rather than device) font in Flash. Reported separately so we learn
        // WHICH of the two works, not just that one of them did.
        t.embedFonts = true;
        t.setTextFormat(fmt);
        r = r + "twe=" + Math.round(t.textWidth) + ";";
        // and whether autoSize/wordWrap are honoured on a runtime field
        t.wordWrap = true;
        t.multiline = true;
        r = r + "th=" + Math.round(t.textHeight) + ";";
    }

    // --- 5. control: an AUTHORED field, same font ---------------------------
    // Isolates "runtime fields can't reach the font" from "no font at all".
    if (root.ctl == undefined) {
        r = r + "atw=nil;";
    } else {
        r = r + "atw=" + Math.round(root.ctl.textWidth) + ";";
    }

    // --- 6. attachMovie of an exported symbol -------------------------------
    var am = root.attachMovie("ProbeRow", "probeRow0", 520);
    r = r + "am=" + (am == undefined ? "0" : "1") + ";";
    // does the attached clip accept a runtime event handler? (mouse support)
    if (root.probeRow0 != undefined) {
        root.probeRow0.onRelease = function () { fscommand("essProbeClick", "row0"); };
        r = r + "evt=" + (root.probeRow0.onRelease == undefined ? "0" : "1") + ";";
    }

    // --- 7. depth allocation + teardown ------------------------------------
    r = r + "nhd=" + root.getNextHighestDepth() + ";";
    holder.removeMovieClip();
    r = r + "rm=" + (root.probeHolder == undefined ? "1" : "0") + ";";

    fscommand("essProbe", r);
    return r;
}

// A second entry point so the host can confirm Lua -> movie calls work at all,
// independently of whether any of the runtime API does.
function Ping() {
    fscommand("essProbe", "ping=1;");
    return 1;
}

// Lets the host write into the on-screen status field. There is no SetVariable
// native on FlashWidget, so a bound variable can only be reached through an AS2
// setter like this one.
function SetStatus(s) {
    _root.probe_status = s;
}
`;

// ------------------------------------------------------------------ the movie
const m = new GFMovie.Movie(320, 200, {
  name: 'ess_probe',
  fps: 30,
  background: [16, 17, 21],
});

// A visible frame so there's something on screen if anyone does look, and a
// baseline that the movie itself loads at all.
m.rect(0, 0, 320, 22, [232, 140, 24]);
m.text(6, 3, 'ESS PROBE', { size: 13, color: [25, 25, 25] });

// The font CONTROL: an authored field with an instance name, so the probe can
// read its textWidth. Also what pulls the ImportAssets2 font tag into the movie
// — without at least one authored text item there'd be no font to resolve.
m.text(6, 30, 'PROBE0123', { size: 13, color: [225, 229, 236], name: 'ctl', width: 200 });

// A status field the host can write into, so a human watching the screen sees
// the same answer the log gets.
m.text(6, 52, '(waiting)', { size: 11, color: [150, 156, 168], varName: 'probe_status', width: 300 });

// attachMovie needs an exported symbol. A clip placed off-stage keeps it out of
// the way while still defining the sprite that ProbeRow names.
m.clip('rowSrc', -400, -400, 100, 16, [52, 58, 68]).exportAs('ProbeRow');

m.script(Compiler.compileSource(AS2));

const out = m.build();
const dest = path.join(__dirname, 'ess_probe.gfx');
fs.writeFileSync(dest, out);

// Verify before shipping it into a game launch — a malformed movie would waste
// the whole round trip.
const summary = Verify.verifyGfx(out, { functions: ['RunProbe', 'Ping'] });
console.log('wrote ' + dest + '  (' + out.length + ' bytes)');
console.log('functions: ' + summary.functions.join(', '));
console.log('tags: ' + JSON.stringify(summary.tags));

File: examples/mercs2/ess_probe.lua
-- Phase-0 spike driver for the Ess.UI rework. Pairs with ess_probe.js/.gfx.
--
-- Not an OnLoad/OnKey script: it's designed to be sent into the live game in
-- pieces through tools/lua_repl.py, so each stage's result can be read back
-- before the next one runs. That matters because the movie loads a frame late —
-- separate REPL calls give it time to be ready without needing a timer.
--
-- SHARED CONTRACT with ess_probe.js:
--   asset          "ess_probe"     -> SetSwfFile("ess_probe.gfx")
--   Lua -> movie   RunProbe()      runs every check
--   Lua -> movie   Ping()          proves the call channel alone
--   movie -> Lua   fscommand("essProbe", "<k=v;...>")
--   movie -> Lua   fscommand("essProbeClick", "row0")   (runtime onRelease)
--
-- Usage, one lua_repl call per stage:
--   1) EssProbe.start()    build the widget + register the handlers
--   2) EssProbe.ping()     confirm Lua -> movie works at all
--   3) EssProbe.run()      run the capability checks
--   4) EssProbe.result()   read the reported line back

import("MrxGuiBase")
import("MrxGuiManager")

_G.EssProbe = _G.EssProbe or {}
local P = _G.EssProbe

P.report = P.report or "(none)"
P.clicked = P.clicked or "(none)"
P.events = P.events or 0

-- Stage 1. Build the widget and register both event handlers BEFORE anything is
-- called into the movie, so no result can be missed.
function P.start()
    local okp, player = pcall(Player.GetLocalPlayer)
    local ok, w = pcall(function()
        local wg = MrxGuiBase.FlashWidget:new()
        if okp and player then pcall(function() wg:SetOwner(player) end) end
        -- SetLocation takes CORNER coords (x1,y1,x2,y2), not (x,y,w,h) -- the
        -- bug Ess.Gfx.widget exists to make unrepeatable.
        wg:SetLocation(40, 60, 40 + 320, 60 + 200)
        wg:SetSwfFile("ess_probe.gfx", nil, nil)
        MrxGuiBase.AddWidget(wg)
        if okp and player then pcall(function() MrxGuiManager.AddWidgetToHud(player, wg) end) end
        return wg
    end)
    if not ok or not w then return "widget FAILED" end
    P.w = w

    -- The movie -> Lua channel. This whole direction is unproven in this kit
    -- (nothing in uilib uses fscommand), so it's being tested here too, not
    -- assumed.
    pcall(function()
        w:SetFlashEventHandler("essProbe", function(_, v)
            P.report = tostring(v)
            P.events = P.events + 1
            Loader.Printf("[essprobe] report: " .. tostring(v))
        end, {})
    end)
    pcall(function()
        w:SetFlashEventHandler("essProbeClick", function(_, v)
            P.clicked = tostring(v)
            Loader.Printf("[essprobe] click: " .. tostring(v))
        end, {})
    end)

    pcall(function() w:SetVisible(true) end)
    return "widget ok"
end

local function call(fn)
    if not P.w then return false end
    return (pcall(function() P.w:CallActionScriptCallback(fn, {}) end)) and true or false
end

-- Stage 2. Proves Lua -> movie calls land at all. If the report comes back
-- "ping=1;" then the call channel AND the fscommand channel both work, which
-- separates a movie-side failure from a plumbing failure in stage 3.
function P.ping()
    P.report = "(none)"
    local ok = call("Ping")
    return "ping sent=" .. tostring(ok)
end

-- Stage 3. The capability checks.
function P.run()
    P.report = "(none)"
    local ok = call("RunProbe")
    return "run sent=" .. tostring(ok)
end

-- Stage 4. Read back what the movie said. Single line by construction, because
-- lua_repl truncates at the first newline.
function P.result()
    return tostring(P.report) .. "  [events=" .. tostring(P.events) .. " click=" .. tostring(P.clicked) .. "]"
end

-- Also write into the movie's own status field, so anyone watching the screen
-- sees the same answer the log gets.
function P.echo(s)
    if not P.w then return false end
    return (pcall(function() P.w:CallActionScriptCallback("SetStatus", { tostring(s) }) end)) and true or false
end

function P.teardown()
    if P.w then pcall(function() P.w:SetVisible(false) end) end
    P.w = nil
    return "torn down"
end

return "EssProbe loaded"
